mogp_gpu/src/gppriors.hpp

```diff
@@ -0,0 +1,477 @@
+
+        check_theta(theta);
+
+        vec partials(theta.get_n_data());
+        //correlation length parameters
+        for (int i=0; i< corr_priors.size(); ++i) {
+            partials[i] = corr_priors[i]->dlogpdtheta(theta.get_corr()[i],CorrTransform());   
+        }
+        // covariance parameter
+        partials[theta.get_n_corr()] = cov_prior->dlogpdtheta(theta.get_cov(), CovTransform());
+        // nugget parameter (if fitted nugget)
+        if (nug_type == NUG_FIT) {
+        //    partials.push_back(nug_prior->dlogpdtheta(theta.get_nugget_size(), CovTransform()));
+            partials[theta.get_n_data()-1] = nug_prior->dlogpdtheta(theta.get_nugget_size(), CovTransform());  
+        }
+        return partials;
+    }
+
+    vec d2logpdtheta2(GPParams theta) {
+
+        check_theta(theta);
+
+        vec hessian(theta.get_n_data());
+        //correlation length parameters
+        for (int i=0; i< corr_priors.size(); ++i) {
+            hessian[i] = corr_priors[i]->d2logpdtheta2(theta.get_corr()[i],CorrTransform());
+        }
+        // covariance parameter
+        hessian[theta.get_n_corr()] = cov_prior->d2logpdtheta2(theta.get_cov(), CovTransform());
+        // nugget parameter (if fitted nugget)
+        if (nug_type == NUG_FIT) {
+            hessian[theta.get_n_data()-1] = nug_prior->d2logpdtheta2(theta.get_nugget_size(), CovTransform()); 
+        }
+        return hessian;
+    }
+
+    std::vector<REAL> sample() {
+        std::vector<REAL> samples; 
+        if ( mean_prior != NULL) {
+            samples = mean_prior->sample(CorrTransform());
+        } 
+        for (int i=0; i< corr_priors.size(); ++i) {
+            samples.push_back(corr_priors[i]->sample(CorrTransform()));
+        }  
+        samples.push_back(cov_prior->sample(CovTransform()));
+        if (nug_type == NUG_FIT) {
+            samples.push_back(nug_prior->sample(CovTransform())); 
+        }
+        return samples;
+    }
+
+    WeakPrior* make_prior(prior_type ptype, REAL param_1, REAL param_2) { 
+
+        if (ptype == INVGAMMA) {
+            return new InvGammaPrior(param_1, param_2); // shape and scale
+        } else if (ptype == GAMMA) {
+            return new GammaPrior(param_1, param_2); // shape and scale
+        } else if (ptype == LOGNORMAL) {
+            return new LogNormalPrior(param_1, param_2); // shape and scale
+        } else if (ptype == WEAK) {
+            return new WeakPrior();
+        } else {
+            std::cout<<"Non-weak Prior must be Inverse Gamma, Gamma, or LogNormal"<<std::endl;  
+            return new WeakPrior();
+        }
+    }
+
+    void create_corr_priors(prior_type ptype, REAL param_1, REAL param_2) {
+        for (int i=0; i< n_corr; ++i) {
+            WeakPrior* new_prior = make_prior(ptype, param_1, param_2);
+            corr_priors.push_back(new_prior);
+        }
+    }
+
+    void create_cov_prior(prior_type ptype, REAL param_1, REAL param_2) {
```

Comment:
Why is there no nugget version of this? Is it because the nugget doesn't always have a prior?

Reply:
Perhaps it could be better named to be in common with these, but `set_nugget` does the same job for the nugget prior if nug_type is "fit".